Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrType.kt

```diff
@@ -0,0 +1,133 @@
+
+private fun KotlinClassifierType.toStubIrType(): StubType =
+        if (PredefinedTypesHandler.isPredefinedType(this)) {
+            PredefinedTypesHandler.expandPrimitiveVarType(this)
+        } else {
+            val typeArguments = arguments.map { it.toStubIrType() }
+            ClassifierStubType(classifier, typeArguments, nullable)
+        }
+
+private fun KotlinTypeArgument.toStubIrType(): TypeArgument = when (this) {
+    is KotlinType -> TypeArgumentStub(this.toStubIrType())
+    StarProjection -> TypeArgument.StarProjection
+    else -> error("Unexpected KotlinTypeArgument: $this")
+}
+
+/**
+ * Types that come from kotlinx.cinterop require special handling because we
+ * don't have explicit information about their structure.
+ * For example, to be able to produce metadata-based interop library we need to know
+ * that ByteVar is a typealias to ByteVarOf<Byte>.
+ */
+private object PredefinedTypesHandler {
+    private const val cInteropPackage = "kotlinx.cinterop"
+
+    private val primitives = setOf(
+            KotlinTypes.boolean,
+            KotlinTypes.byte, KotlinTypes.short, KotlinTypes.int, KotlinTypes.long,
+            KotlinTypes.uByte, KotlinTypes.uShort, KotlinTypes.uInt, KotlinTypes.uLong,
+            KotlinTypes.float, KotlinTypes.double
+    )
+
+    /**
+     * kotlinx.cinterop.{primitive}Var -> kotlin.{primitive}
+     */
+    private val primitiveVarClassifierToPrimitiveType: Map<Classifier, KotlinClassifierType> =
+            primitives.associateBy {
+                val typeVar = "${it.classifier.topLevelName}Var"
+                Classifier.topLevel(cInteropPackage, typeVar)
+            }
+
+    /**
+     * @param primitiveType primitive type from kotlin package.
+     * @return kotlinx.cinterop.[primitiveType]VarOf<[primitiveType]>
+     */
+    private fun getVarOfTypeFor(primitiveType: KotlinClassifierType): ClassifierStubType {
+        val typeVarOf = "${primitiveType.classifier.topLevelName}VarOf"
+        val classifier = Classifier.topLevel(cInteropPackage, typeVarOf)
+        return ClassifierStubType(classifier, listOf(TypeArgumentStub(primitiveType.toStubIrType())))
+    }
+
+    /**
+     * @param primitiveVarType one of kotlinx.cinterop.{primitive}Var types.
+     * @return typealias in terms of StubIR types.
+     */
+    fun expandPrimitiveVarType(primitiveVarType: KotlinClassifierType): AbbreviationStubType {
+        val primitiveType = primitiveVarClassifierToPrimitiveType.getValue(primitiveVarType.classifier)
+        val expandedType = getVarOfTypeFor(primitiveType)
+        val abbreviation = ClassifierStubType(primitiveVarType.classifier)
+        return AbbreviationStubType(expandedType, abbreviation)
+    }
+
+    /**
+     * Check that provided type should be handled by [expandPrimitiveVarType] function.
+     */
+    fun isPredefinedType(kotlinClassifierType: KotlinClassifierType): Boolean =
```

Comment:
Why not make one safe `AbbreviationStubType?`-returning function instead of making predicate and function that expects it to be `true`?